src/scala/microsoft-spark-2.4.x/src/main/scala/org/apache/spark/api/dotnet/CallbackConnection.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the .NET Foundation under one or more agreements.
+ * The .NET Foundation licenses this file to you under the MIT license.
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.apache.spark.api.dotnet
+
+import java.io.{ByteArrayOutputStream, Closeable, DataInputStream, DataOutputStream}
+import java.net.Socket
+
+import org.apache.spark.internal.Logging
+
+/**
+ * CallbackConnection is used to process the callback communication
+ * between the JVM and Dotnet. It uses a TCP socket to communicate with
+ * the Dotnet CallbackServer and the socket is expected to be reused.
+ * @param address The address of the Dotnet CallbackServer
+ * @param port The port of the Dotnet CallbackServer
+ */
+class CallbackConnection(address: String, port: Int) extends Logging {
+  private[this] val socket: Socket = new Socket(address, port)
+  private[this] val inputStream: DataInputStream = new DataInputStream(socket.getInputStream)
+  private[this] val outputStream: DataOutputStream = new DataOutputStream(socket.getOutputStream)
+
+  def send(
+      callbackId: Int,
+      writeBody: DataOutputStream => Unit): ConnectionStatus.ConnectionStatus = {
+    logInfo(s"Calling callback [callback id = $callbackId] ...")
+
+    try {
+      SerDe.writeInt(outputStream, CallbackFlags.CALLBACK)
+      SerDe.writeInt(outputStream, callbackId)
+
+      val byteArrayOutputStream = new ByteArrayOutputStream()
+      writeBody(new DataOutputStream(byteArrayOutputStream))
+      SerDe.writeInt(outputStream, byteArrayOutputStream.size)
+      byteArrayOutputStream.writeTo(outputStream);
+    } catch {
+      case e: Exception => {
+        logError("Error writing to stream.", e)
+        return ConnectionStatus.ERROR_WRITE
+      }
+    }
+
+    logInfo(s"Signaling END_OF_STREAM.")
+    try {
+      SerDe.writeInt(outputStream, CallbackFlags.END_OF_STREAM)
+      outputStream.flush()
+
+      val endOfStreamResponse = readFlag(inputStream)
+      endOfStreamResponse match {
+        case CallbackFlags.END_OF_STREAM =>
+          logInfo(s"Received END_OF_STREAM signal. Calling callback [callback id = $callbackId] successful.")
+          return ConnectionStatus.SUCCESS
+        case _ =>  {
+          logError(s"Error verifying end of stream. Expected: ${CallbackFlags.END_OF_STREAM}, " +
+              s"Received: $endOfStreamResponse")
+        }
+      }
+    } catch {
+      case e: Exception => {
+        logError("Error while verifying end of stream.", e)
+      }
+    }
+
+    ConnectionStatus.ERROR_END_OF_STREAM
```

Comment:
Now that we don't care about the status, should we just throw with the right exception message to simplify further?

Reply:
that means it'll be more changes if we decide to add retries in the future.  Ok with you ?

Reply:
At least we will have a code to reference to? And it will be smaller than this PR :)

Reply:
throwing exceptions instead of status codes.